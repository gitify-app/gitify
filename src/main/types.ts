export interface GitifyAPI {
  openExternalLink: (url: string) => void;
  getAppVersion: () => Promise<string>;
  encryptValue: (value: string) => Promise<string>;
  decryptValue: (value: string) => Promise<string>;
  quitApp: () => void;
  showWindow: () => void;
  hideWindow: () => void;
  setAutoLaunch: (value: boolean) => void;
  setAlternateIdleIcon: (value: boolean) => void;
  setKeyboardShortcut: (keyboardShortcut: boolean) => void;
  updateTrayIcon: (notificationsLength?: number) => void;
  updateTrayTitle: (title?: string) => void;
}
