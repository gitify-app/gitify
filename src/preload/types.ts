import type { api } from '.';

export type GitifyAPI = typeof api;
