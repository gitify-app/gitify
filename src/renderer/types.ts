import type { OcticonProps } from '@primer/octicons-react';
import type { FC } from 'react';
import type { Notification, Reason } from './typesGitHub';
import type {
  AuthMethod,
  EnterpriseAccount,
  PlatformType,
} from './utils/auth/types';

declare const __brand: unique symbol;

type Brand<B> = { [__brand]: B };

export interface AuthState {
  accounts: Account[];
  /**
   * @deprecated This attribute is deprecated and will be removed in a future release.
   */
  token?: Token;
  /**
   * @deprecated This attribute is deprecated and will be removed in a future release.
   */
  enterpriseAccounts?: EnterpriseAccount[];
  /**
   * @deprecated This attribute is deprecated and will be removed in a future release.
   */
  user?: GitifyUser | null;
}

export type Branded<T, B> = T & Brand<B>;

export type AuthCode = Branded<string, 'AuthCode'>;

export type Token = Branded<string, 'Token'>;

export type ClientID = Branded<string, 'ClientID'>;

export type ClientSecret = Branded<string, 'ClientSecret'>;

export type Hostname = Branded<string, 'Hostname'>;

export type Link = Branded<string, 'WebUrl'>;

export type Status = 'loading' | 'success' | 'error';

export interface Account {
  method: AuthMethod;
  platform: PlatformType;
  version?: string;
  hostname: Hostname;
  token: Token;
  user: GitifyUser | null;
  hasRequiredScopes?: boolean;
}

export type SettingsValue =
  | boolean
  | number
  | GroupBy
  | OpenPreference
  | Reason[]
  | Theme;

export type SettingsState = AppearanceSettingsState &
  NotificationSettingsState &
  SystemSettingsState &
  FilterSettingsState;

export interface AppearanceSettingsState {
  theme: Theme;
  zoomPercentage: number;
  detailedNotifications: boolean;
  showAccountHeader: boolean;
  showPills: boolean;
  showNumber: boolean;
  wrapNotificationTitle: boolean;
}

export interface NotificationSettingsState {
  groupBy: GroupBy;
  fetchAllNotifications: boolean;
  participating: boolean;
  markAsDoneOnOpen: boolean;
  markAsDoneOnUnsubscribe: boolean;
  delayNotificationState: boolean;
}

export interface SystemSettingsState {
  openLinks: OpenPreference;
  keyboardShortcut: boolean;
  showNotificationsCountInTray: boolean;
  showNotifications: boolean;
  useAlternateIdleIcon: boolean;
  playSound: boolean;
  openAtStartup: boolean;
}

export interface FilterSettingsState {
  hideBots: boolean;
  filterReasons: Reason[];
}

export interface GitifyState {
  auth?: AuthState;
  settings?: SettingsState;
}

export enum Theme {
  SYSTEM = 'SYSTEM',
  LIGHT = 'LIGHT',
  LIGHT_HIGH_CONTRAST = 'LIGHT_HIGH_CONTRAST',
  LIGHT_COLORBLIND = 'LIGHT_COLORBLIND',
  LIGHT_TRITANOPIA = 'LIGHT_TRITANOPIA',
  DARK = 'DARK',
  DARK_HIGH_CONTRAST = 'DARK_HIGH_CONTRAST',
  DARK_COLORBLIND = 'DARK_COLORBLIND',
  DARK_TRITANOPIA = 'DARK_TRITANOPIA',
  DARK_DIMMED = 'DARK_DIMMED',
}

export enum OpenPreference {
  FOREGROUND = 'FOREGROUND',
  BACKGROUND = 'BACKGROUND',
}

export enum GroupBy {
  REPOSITORY = 'REPOSITORY',
  DATE = 'DATE',
}

export type RadioGroupItem = {
  label: string;
  value: string;
};

export interface AccountNotifications {
  account: Account;
  notifications: Notification[];
  error: GitifyError | null;
}

export interface GitifyUser {
  login: string;
  name: string | null;
  avatar: Link | null;
  id: number;
}

export interface GitifyError {
  title: string;
  descriptions: string[];
  emojis: string[];
}

export type ErrorType =
  | 'BAD_CREDENTIALS'
  | 'MISSING_SCOPES'
  | 'NETWORK'
  | 'RATE_LIMITED'
  | 'UNKNOWN';

export interface FormattedReason {
  title: string;
  description: string;
}

export enum IconColor {
  GRAY = 'text-gitify-icon-muted',
  GREEN = 'text-gitify-icon-open',
  PURPLE = 'text-gitify-icon-done',
  RED = 'text-gitify-icon-closed',
  YELLOW = 'text-gitify-icon-attention',
}

export enum Opacity {
  READ = 'opacity-50',
  LOW = 'opacity-70',
  MEDIUM = 'opacity-80',
  HIGH = 'opacity-90',
}

export type PullRequestApprovalIcon = {
  type: FC<OcticonProps>;
  color: IconColor;
  description: string;
};

export enum Size {
  XSMALL = 12,
  SMALL = 14,
  MEDIUM = 16,
  LARGE = 18,
  XLARGE = 20,
}

export type Chevron = {
  icon: FC<OcticonProps>;
  label: string;
};
