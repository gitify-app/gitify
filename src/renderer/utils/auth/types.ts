import type {
  AuthCode,
  ClientID,
  ClientSecret,
  Hostname,
  Token,
} from '../../types';

export type AuthMethod = 'GitHub App' | 'Personal Access Token' | 'OAuth App';

export type PlatformType = 'GitHub Cloud' | 'GitHub Enterprise Server';

export interface LoginOAuthAppOptions {
  hostname: Hostname;
  clientId: ClientID;
  clientSecret: ClientSecret;
}

export interface LoginPersonalAccessTokenOptions {
  hostname: Hostname;
  token: Token;
}

export interface AuthResponse {
  authMethod: AuthMethod;
  authCode: AuthCode;
  authOptions: LoginOAuthAppOptions;
}

export interface AuthTokenResponse {
  hostname: Hostname;
  token: Token;
}
