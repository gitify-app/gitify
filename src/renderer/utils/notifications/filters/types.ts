import type {
  AccountNotifications,
  SettingsState,
  TypeDetails,
} from '../../../types';
import type { Notification } from '../../../typesGitHub';

export interface Filter<T extends string> {
  FILTER_TYPES: Record<T, TypeDetails>;

  requiresDetailsNotifications: boolean;

  getTypeDetails(type: T): TypeDetails;

  hasFilters(settings: SettingsState): boolean;

  isFilterSet(settings: SettingsState, type: T): boolean;

  getFilterCount(notifications: AccountNotifications[], type: T): number;

  filterNotification(notification: Notification, type: T): boolean;
}
