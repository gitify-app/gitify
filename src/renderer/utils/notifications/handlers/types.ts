import type { FC } from 'react';

import type { OcticonProps } from '@primer/octicons-react';

import type { SettingsState } from '../../../types';
import type {
  GitifySubject,
  Notification,
  Subject,
  SubjectType,
} from '../../../typesGitHub';

export interface NotificationTypeHandler {
  readonly type?: SubjectType;

  /**
   * Enrich a notification. Settings may be unused for some handlers.
   */
  enrich(
    notification: Notification,
    settings: SettingsState,
  ): Promise<GitifySubject>;

  /** Return an icon component for this notification type. */
  getIcon(subject: Subject): FC<OcticonProps> | null;
}
