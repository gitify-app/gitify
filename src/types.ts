import type { Notification } from './typesGithub';

export interface AuthState {
  token?: string;
  enterpriseAccounts: EnterpriseAccount[];
  user: GitifyUser | null;
}

export type SettingsState = AppearanceSettingsState &
  NotificationSettingsState &
  SystemSettingsState;

interface AppearanceSettingsState {
  theme: Theme;
  detailedNotifications: boolean;
  showAccountHostname: boolean;
}

interface NotificationSettingsState {
  participating: boolean;
  showNotifications: boolean;
  showBots: boolean;
  markAsDoneOnOpen: boolean;
}

interface SystemSettingsState {
  playSound: boolean;
  openAtStartup: boolean;
  showNotificationsCountInTray: boolean;
}

export enum Theme {
  SYSTEM = 'SYSTEM',
  LIGHT = 'LIGHT',
  DARK = 'DARK',
}

export type RadioGroupItem = {
  label: string;
  value: string;
};

export interface EnterpriseAccount {
  hostname: string;
  token: string;
}

export interface AccountNotifications {
  hostname: string;
  notifications: Notification[];
}

export interface AuthOptions {
  hostname: string;
  clientId: string;
  clientSecret: string;
}

export interface AuthTokenOptions {
  hostname: string;
  token: string;
}

export interface AuthResponse {
  authCode: string;
  authOptions: AuthOptions;
}
export interface AuthTokenResponse {
  hostname: string;
  token: string;
}

export interface GitifyUser {
  login: string;
  name: string;
  id: number;
}

export interface GitifyError {
  title: string;
  descriptions: string[];
  emojis: string[];
}

export type ErrorType =
  | 'BAD_CREDENTIALS'
  | 'MISSING_SCOPES'
  | 'NETWORK'
  | 'RATE_LIMITED'
  | 'UNKNOWN';

export interface FormattedReason {
  title: string;
  description: string;
}
