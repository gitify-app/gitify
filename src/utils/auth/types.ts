import type {
  AuthCode,
  ClientID,
  ClientSecret,
  Hostname,
  Token,
} from '../../types';

export type AuthMethod = 'GitHub App' | 'Personal Access Token' | 'OAuth App';

export type PlatformType = 'GitHub Cloud' | 'GitHub Enterprise Server';

export interface LoginOAuthAppOptions {
  hostname: Hostname;
  clientId: ClientID;
  clientSecret: ClientSecret;
}

export interface LoginPersonalAccessTokenOptions {
  hostname: Hostname;
  token: Token;
}

export interface AuthResponse {
  authCode: AuthCode;
  authOptions: LoginOAuthAppOptions;
}

export interface AuthTokenResponse {
  hostname: Hostname;
  token: Token;
}

/**
 * @deprecated This type is deprecated and will be removed in a future release.
 */
export interface EnterpriseAccount {
  hostname: Hostname;
  token: Token;
}
