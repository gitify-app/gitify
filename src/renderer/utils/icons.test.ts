import {
  CheckIcon,
  CommentIcon,
  FeedPersonIcon,
  FileDiffIcon,
  MarkGithubIcon,
  OrganizationIcon,
} from '@primer/octicons-react';

import { IconColor } from '../types';
import type { GitifyPullRequestReview } from '../typesGitHub';
import {
  getAuthMethodIcon,
  getDefaultUserIcon,
  getPlatformIcon,
  getPullRequestReviewIcon,
} from './icons';

describe('renderer/utils/icons.ts', () => {
  describe('getPullRequestReviewIcon', () => {
    let mockReviewSingleReviewer: GitifyPullRequestReview;
    let mockReviewMultipleReviewer: GitifyPullRequestReview;

    beforeEach(() => {
      mockReviewSingleReviewer = {
        state: 'APPROVED',
        users: ['user1'],
      };
      mockReviewMultipleReviewer = {
        state: 'APPROVED',
        users: ['user1', 'user2'],
      };
    });

    it('approved', () => {
      mockReviewSingleReviewer.state = 'APPROVED';
      mockReviewMultipleReviewer.state = 'APPROVED';

      expect(getPullRequestReviewIcon(mockReviewSingleReviewer)).toEqual({
        type: CheckIcon,
        color: IconColor.GREEN,
        description: 'user1 approved these changes',
      });

      expect(getPullRequestReviewIcon(mockReviewMultipleReviewer)).toEqual({
        type: CheckIcon,
        color: IconColor.GREEN,
        description: 'user1, user2 approved these changes',
      });
    });

    it('changes requested', () => {
      mockReviewSingleReviewer.state = 'CHANGES_REQUESTED';
      mockReviewMultipleReviewer.state = 'CHANGES_REQUESTED';

      expect(getPullRequestReviewIcon(mockReviewSingleReviewer)).toEqual({
        type: FileDiffIcon,
        color: IconColor.RED,
        description: 'user1 requested changes',
      });

      expect(getPullRequestReviewIcon(mockReviewMultipleReviewer)).toEqual({
        type: FileDiffIcon,
        color: IconColor.RED,
        description: 'user1, user2 requested changes',
      });
    });

    it('commented', () => {
      mockReviewSingleReviewer.state = 'COMMENTED';
      mockReviewMultipleReviewer.state = 'COMMENTED';

      expect(getPullRequestReviewIcon(mockReviewSingleReviewer)).toEqual({
        type: CommentIcon,
        color: IconColor.YELLOW,
        description: 'user1 left review comments',
      });

      expect(getPullRequestReviewIcon(mockReviewMultipleReviewer)).toEqual({
        type: CommentIcon,
        color: IconColor.YELLOW,
        description: 'user1, user2 left review comments',
      });
    });

    it('dismissed', () => {
      mockReviewSingleReviewer.state = 'DISMISSED';
      mockReviewMultipleReviewer.state = 'DISMISSED';

      expect(getPullRequestReviewIcon(mockReviewSingleReviewer)).toEqual({
        type: CommentIcon,
        color: IconColor.GRAY,
        description: 'user1 review has been dismissed',
      });

      expect(getPullRequestReviewIcon(mockReviewMultipleReviewer)).toEqual({
        type: CommentIcon,
        color: IconColor.GRAY,
        description: 'user1, user2 reviews have been dismissed',
      });
    });

    it('pending', () => {
      mockReviewSingleReviewer.state = 'PENDING';
      mockReviewMultipleReviewer.state = 'PENDING';

      expect(getPullRequestReviewIcon(mockReviewSingleReviewer)).toBeNull();

      expect(getPullRequestReviewIcon(mockReviewMultipleReviewer)).toBeNull();
    });
  });

  describe('getAuthMethodIcon', () => {
    expect(getAuthMethodIcon('GitHub App')).toMatchSnapshot();

    expect(getAuthMethodIcon('OAuth App')).toMatchSnapshot();

    expect(getAuthMethodIcon('Personal Access Token')).toMatchSnapshot();
  });

  describe('getPlatformIcon', () => {
    expect(getPlatformIcon('GitHub Cloud')).toMatchSnapshot();

    expect(getPlatformIcon('GitHub Enterprise Server')).toMatchSnapshot();
  });

  describe('getDefaultUserIcon', () => {
    expect(getDefaultUserIcon('Bot')).toBe(MarkGithubIcon);
    expect(getDefaultUserIcon('EnterpriseUserAccount')).toBe(FeedPersonIcon);
    expect(getDefaultUserIcon('Mannequin')).toBe(MarkGithubIcon);
    expect(getDefaultUserIcon('Organization')).toBe(OrganizationIcon);
    expect(getDefaultUserIcon('User')).toBe(FeedPersonIcon);
  });
});
