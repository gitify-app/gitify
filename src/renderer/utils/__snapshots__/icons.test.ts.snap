// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
{
  "$$typeof": Symbol(react.forward_ref),
  "render": [Function],
}
`;

exports[` 2`] = `
{
  "$$typeof": Symbol(react.forward_ref),
  "render": [Function],
}
`;

exports[` 3`] = `
{
  "$$typeof": Symbol(react.forward_ref),
  "render": [Function],
}
`;

exports[` 4`] = `
{
  "$$typeof": Symbol(react.forward_ref),
  "render": [Function],
}
`;

exports[` 5`] = `
{
  "$$typeof": Symbol(react.forward_ref),
  "render": [Function],
}
`;

exports[`renderer/utils/icons.ts getNotificationTypeIconColor should format the notification color for check suite 1`] = `"text-primer-muted"`;

exports[`renderer/utils/icons.ts getNotificationTypeIconColor should format the notification color for check suite 2`] = `"text-primer-closed"`;

exports[`renderer/utils/icons.ts getNotificationTypeIconColor should format the notification color for check suite 3`] = `"text-primer-muted"`;

exports[`renderer/utils/icons.ts getNotificationTypeIconColor should format the notification color for check suite 4`] = `"text-primer-open"`;

exports[`renderer/utils/icons.ts getNotificationTypeIconColor should format the notification color for check suite 5`] = `"text-primer-muted"`;

exports[`renderer/utils/icons.ts getNotificationTypeIconColor should format the notification color for state 1`] = `"text-primer-open"`;

exports[`renderer/utils/icons.ts getNotificationTypeIconColor should format the notification color for state 2`] = `"text-primer-closed"`;

exports[`renderer/utils/icons.ts getNotificationTypeIconColor should format the notification color for state 3`] = `"text-primer-done"`;

exports[`renderer/utils/icons.ts getNotificationTypeIconColor should format the notification color for state 4`] = `"text-primer-muted"`;

exports[`renderer/utils/icons.ts getNotificationTypeIconColor should format the notification color for state 5`] = `"text-primer-done"`;

exports[`renderer/utils/icons.ts getNotificationTypeIconColor should format the notification color for state 6`] = `"text-primer-muted"`;

exports[`renderer/utils/icons.ts getNotificationTypeIconColor should format the notification color for state 7`] = `"text-primer-open"`;

exports[`renderer/utils/icons.ts getNotificationTypeIconColor should format the notification color for state 8`] = `"text-primer-open"`;

exports[`renderer/utils/icons.ts getNotificationTypeIconColor should format the notification color for state 9`] = `"text-primer-done"`;

exports[`renderer/utils/icons.ts getNotificationTypeIconColor should format the notification color for state 10`] = `"text-primer-muted"`;
